feat: Add RPC (sub)domain tracking to transaction event metrics for RPC endpoints usage

httpJunkie opened the pull request:
## **Description**

This PR implements GitHub [Issue #2707](https://github.com/MetaMask/MetaMask-planning/issues/2707) by adding the ability to track what type of RPC endpoint users are connecting to. We now capture whether users connect to Ethereum using:

- Infura provider (default)
- Other providers (like tenderly.co or alchemy.com)
- Private IP addresses
- Localhost

The implementation includes:

- A utility function `extractRpcDomain` that parses RPC URLs and returns just the domain part
- A domain categorization system that identifies known vs private endpoints
- Extension of transaction metrics to include an `rpc_domain` property for "Transaction Submitted" and "Transaction Finalized" events
- Integration with the NetworkController to retrieve the correct RPC URL for a given chain
- Cache-only approach to retrieving chain data that respects user privacy settings

This data will allow us to better understand the most popular RPC endpoints and what percentage of users customize them, helping us optimize our UX accordingly.

### **Privacy and Security Considerations**

- Only stores domains from well-known public RPC providers
- Returns "private" for custom/unknown endpoints to protect user privacy
- Uses locally cached chain data instead of making network requests
- Works regardless of the "Network details check" setting (respects user preferences)

### **Implementation Notes**

- The implementation uses `getSafeChainsListFromCacheOnly` to avoid making network requests
- This aligns with upcoming refactoring that will rename "safe chains" to "well-known chains"
- Error handling is included to ensure graceful behavior if chain data isn't available

### **Possible values for `rpc_domain`**
- Public RPC provider domains: e.g., `mainnet.infura.io`, `eth-mainnet.alchemyapi.io`
- Private/Custom endpoints: `private`
- Invalid URLs (if encountered): `invalid`

## **Related issues**
[MetaMask-planning - #2707](https://github.com/MetaMask/MetaMask-planning/issues/2707)

## **Manual testing steps**

Set up MetaMask with different RPC endpoints:

1. Default Infura provider
2. Custom provider (e.g., Alchemy)
3. Private IP/localhost

Submit transactions on these networks and verify that the transaction metrics events include the correct `rpc_domain` property. Additionally, verify behavior with the "Network details check" setting both enabled and disabled to ensure the feature works in both cases.

## **Screenshots/Recordings**
<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**
N/A (Payloads would just not have the `rpc_domain` property.

### **After**

<img width="300" alt="01-transaction-finalized" src="https://github.com/user-attachments/assets/2311c1e3-2e53-4cd1-b686-3b16842eb4e3" />
<img width="300" alt="02- transaction-finalized-anon" src="https://github.com/user-attachments/assets/91cf2b6d-98a8-4fa1-b0ec-1e9701e2a132" />
<img width="300" alt="03-transaction-approved" src="https://github.com/user-attachments/assets/68b0cc1d-1d81-4db3-8813-595c361cc303" />
<img width="300" alt="04-transaction-approved-anon" src="https://github.com/user-attachments/assets/6955c63e-b1c3-4d6f-a58f-aaad5d7ba011" />
<img width="300" alt="05-transaction-submitted" src="https://github.com/user-attachments/assets/9d0b2893-c2ee-40da-8af8-8c3b7c138bd3" />
<img width="300" alt="06-transaction-submitted-anon" src="https://github.com/user-attachments/assets/da4f7e9c-6618-454c-b5f3-641e8d0b1111" />
<img width="300" alt="07-transaction-added_NO-RPCURL" src="https://github.com/user-attachments/assets/26ed6ca3-c146-48cc-a9f6-30dce085c85d" />
<img width="300" alt="08-transaction-added-anon_NO-RPCURL" src="https://github.com/user-attachments/assets/0d33219e-0f19-43c4-9803-824049e22174" />


## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

httpJunkie committed:
feat: placeholder for RPC domain util

initial implementation for GitHub Issue #2707.

Added a placeholder function for RPC domain util. This will be expanded in subsequent commits to handle the various requirements

httpJunkie committed:
test: add extractRpcDomain test placeholder

httpJunkie committed:
feat: fix extractRpcDomain to have actual implementation that satisfies basic reqs
test: fix extractRpcDomain test to have actual implementation that satisfies basic reqs

httpJunkie committed:
Merge branch 'main' into feat/metrics-for-rpc-endpoints

httpJunkie committed:
Merge branch 'main' into feat/metrics-for-rpc-endpoints

httpJunkie committed:
chore: add PR and description to changelog doc

httpJunkie committed:
feat: Add RPC domain tracking to transaction metrics

Implements GitHub Issue #2707 to track what type of RPC endpoint users are connecting to:
- Added getNetworkRpcUrl function to extract RPC URL from network configuration
- Modified transaction metrics to include rpc_domain property
- Domain extraction categorizes endpoints as Infura, other providers, private IPs, or localhost
- Applied to transaction events including Transaction Submitted and Transaction Finalized

This enables analytics on the percentage of users using different RPC endpoint types.

httpJunkie committed:
feat: restrict rpc-url in status values "confirmed" and "submitted"

These should correspond to events:

- "Transaction Finalized"
- "Transaction Submitted"

httpJunkie committed:
Merge branch 'main' into feat/metrics-for-rpc-endpoints

httpJunkie committed:
Merge branch 'main' into feat/metrics-for-rpc-endpoints

httpJunkie committed:
remove restriction for status submitted and confirm

It is fine for the new properties to show on all events that the `buildEventFragmentProperties` sets them up for.

httpJunkie committed:
fix:tests: add restrictions back as it ws causing 16 tests to fail, update one test to expect new property and value.

httpJunkie committed:
fix: make pretty

httpJunkie committed:
Merge branch 'main' into feat/metrics-for-rpc-endpoints

httpJunkie committed:
Merge branch 'feat/metrics-for-rpc-endpoints' of github.com:MetaMask/metamask-extension into feat/metrics-for-rpc-endpoints

httpJunkie committed:
todo: We should move getNetworkRpcUrl to the `core/network-controller`

httpJunkie committed:
Merge branch 'main' into feat/metrics-for-rpc-endpoints

davidmurdoch commented on app/scripts/lib/util.ts:
```diff
@@ -479,3 +479,58 @@ export function getConversionRatesForNativeAsset({
+    // Extract hostname
+    const { hostname } = url;
+
+    // Handle localhost (including standard localhost IPs)
+    if (
+      hostname === 'localhost' ||
+      hostname === '127.0.0.1' ||
+      hostname.startsWith('192.168.')
+    ) {
+      return 'localhost';
+    }
```
If we need to detect all loopback addresses, its the entire `127.*` block (.e., `127.0.0.0/8`). And for ipv6 (which we don't _currently_ support), its `0:0:0:0:0:0:0:0001` (and all of the shorthand variants of it 😅).

Since you are including `192.168` addresses in here, there _are_ additional "local" network addresses as well:

`0.0.0.0 – 10.255.255.255`,
`172.16.0.0 – 172.31.255.255`,
`192.168.0.0 – 192.168.255.255`,
and for IPv6
`fd00::/8`

Almost all of these will get caught by the `private_url` return value, but while their use is less common than what you listed above, they could cause `'private_url'` to be over reported.

davidmurdoch commented on app/scripts/lib/util.ts:
```diff
@@ -479,3 +479,58 @@ export function getConversionRatesForNativeAsset({
+    // Check for private IP addresses using regex
+    const ipRegex = /^(?:[0-9]{1,3}\.){3}[0-9]{1,3}$/u;
+    if (ipRegex.test(hostname)) {
+      return 'private_url';
```
I might argue that these represent `ip_address` and not `private_url`, since `url` is a broad term. I don't know if you can change the name or not, but if you can, it might be nice in the future if it didn't use `url`. just a thought.

httpJunkie replied:
I can certainly check on changing the name. I will bring that up. I will check with those defining the req's

davidmurdoch commented on app/scripts/lib/util.ts:
```diff
@@ -479,3 +479,58 @@ export function getConversionRatesForNativeAsset({
+    }
+
+    // Return the domain (hostname) for all other cases
+    return hostname;
```
This still can return a user's private information.

Simple example: if I set up a node in my house on my own domain, `davidmurdoch.com`, MetaMask would now have a pretty good idea who I am. I actually have friends who run their own RPCs behind their own personal domain name.

It is also technically possible that a provider could embed the user's access ID (like INFURA_ID) in the subdomain (hostname), and we'd be leaking it if we did this.

Initially I was thinking we could hash the `hostname`, but there are only like half a million domains, so it is pretty trivial to reverse all domain hashes to an existing domain.

Maybe we can filter by a known list of RPC providers instead?

Or some other ideas?

httpJunkie committed:
Merge branch 'main' into feat/metrics-for-rpc-endpoints

httpJunkie committed:
Merge branch 'main' into feat/metrics-for-rpc-endpoints

httpJunkie committed:
Merge branch 'main' into feat/metrics-for-rpc-endpoints

httpJunkie commented:
I have read the CLA Document and I hereby sign the CLA

httpJunkie committed:
Refactor RPC domain tracking for transaction metrics

Improve network type detection
Enhance privacy protection
Use consistent naming conventions

httpJunkie committed:
feat: update the extractRpcDomain function to meet the knownProvider ? private : invalid` new requirement.

httpJunkie committed:
fix: rename `initializeKnownDomains` to `initializeRpcProviderDomains` and initialize early in app lifecycle within `initializeChainlist` an existing "async" function.

httpJunkie committed:
feat: Add RPC domain extraction for transaction metrics

- Implement `extractRpcDomain` function to extract domains from RPC endpoints
- Return known provider domains (e.g. infura.io) as-is for metrics
- Return 'private' for local, private IP, or custom domains to protect user privacy
- Return 'invalid' for missing or malformed URLs
- Prepare for integration with chain list data for known providers
- Add comprehensive test coverage for all scenarios

This implementation supports tracking RPC provider domains in transaction metrics
while ensuring user privacy for custom or private endpoints.

httpJunkie committed:
fix: improve extractRpcDomain to handle whitespace and IP edge cases

- Trim input URL to prevent false negatives from leading/trailing spaces
- Improved handling of IPv4 and IPv6 addresses using stricter regex
- Fixed test cases to ensure knownDomains are passed when expected
- Added support for URLs without protocol while preserving domain parsing
- Improved error handling for malformed URLs

httpJunkie committed:
Merge branch 'main' into feat/metrics-for-rpc-endpoints

davidmurdoch commented on app/scripts/lib/util.ts:
```diff
@@ -480,3 +481,138 @@ export function getConversionRatesForNativeAsset({
+
+    if (
+      hostname === 'localhost' ||
+      hostname === '::1' ||
+      hostname.startsWith('127.')
+    ) {
+      return 'private';
+    }
+
+    const validIPv4 =
+      /^(?:(?:25[0-5]|2[0-4]\d|1?\d{1,2})\.){3}(?:25[0-5]|2[0-4]\d|1?\d{1,2})$/u;
+
+    const isPrivateIPv4 = (ip: string): boolean => {
+      if (!validIPv4.test(ip)) {
+        return false;
+      }
+      const [a, b] = ip.split('.').map(Number);
+      return (
+        a === 10 ||
+        (a === 172 && b >= 16 && b <= 31) ||
+        (a === 192 && b === 168)
+      );
+    };
+
+    if (validIPv4.test(hostname) && isPrivateIPv4(hostname)) {
+      return 'private';
+    }
+
+    if (hostname.includes(':')) {
+      return 'private';
+    }
+
+    if (hostname.startsWith('xn--')) {
+      return 'private';
+    }
+
+    const domainsToCheck = knownDomainsForTesting || knownDomainsSet;
+    if (domainsToCheck?.has(hostname)) {
+      return hostname;
+    }
```
If you put this check _first_ I think you can get rid of the rest of the logic and return `private` in the `else` case. Previously the code was trying to differentiate between different types of urls, but now that we only have two it doesn't make much sense to do the more complicated logic above.

httpJunkie replied:
Yeah I rushed that change. Sorry. I got it updated and ran the tests.

davidmurdoch commented on app/scripts/lib/util.ts:
```diff
@@ -480,3 +481,138 @@ export function getConversionRatesForNativeAsset({
+    try {
+      url = new URL(trimmedRpcUrl);
+    } catch (e) {
+      if (
+        !trimmedRpcUrl.startsWith('http://') &&
+        !trimmedRpcUrl.startsWith('https://')
+      ) {
+        try {
+          url = new URL(`https://${trimmedRpcUrl}`);
+        } catch (e2) {
+          return 'invalid';
+        }
```
nit: I don't think this error handling logic will ever be used in practice, since our interface and API will always require a complete and valid URL. If it wasn't valid, this code could never run anyway, since it'd fail earlier. I think you can just return `invalid` right away.

httpJunkie replied:
yes I was thinking the same after I committed it. Came back and tested it out. Your right.

httpJunkie replied:
I've also removed some tests that made this assumption and knowing the API will always require a complete and valid URL, I've tried to remove anything that is uneeeded

davidmurdoch commented on app/scripts/lib/util.ts:
```diff
@@ -480,3 +481,138 @@ export function getConversionRatesForNativeAsset({
+    return 'invalid';
+  }
+
+  const trimmedRpcUrl = rpcUrl.trim();
```
`new URL(...)` trims the string on its own.

httpJunkie replied:
I also tested this and confirmed. TY

davidmurdoch commented on app/scripts/metamask-controller.js:
```diff
@@ -7798,6 +7802,50 @@ export default class MetamaskController extends EventEmitter {
+          // Find the network client ID for this chain
+          const networkClientId =
+            this.networkController.findNetworkClientIdByChainId(chainId);
+
+          if (networkClientId) {
+            // Get the network configuration
+            const networkConfig =
+              this.networkController.getNetworkConfigurationByNetworkClientId(
+                networkClientId,
+              );
+
+            if (networkConfig) {
+              // Try direct rpcUrl property first
+              if (networkConfig.rpcUrl) {
+                return networkConfig.rpcUrl;
+              }
+
+              // Try rpcEndpoints array
+              if (
+                networkConfig.rpcEndpoints &&
+                networkConfig.rpcEndpoints.length > 0
+              ) {
+                // Use the default endpoint if specified
+                const defaultEndpointIndex =
+                  networkConfig.defaultRpcEndpointIndex || 0;
+                if (networkConfig.rpcEndpoints[defaultEndpointIndex]) {
+                  return networkConfig.rpcEndpoints[defaultEndpointIndex].url;
+                }
+
+                // Fallback to first endpoint
+                if (networkConfig.rpcEndpoints[0].url) {
+                  return networkConfig.rpcEndpoints[0].url;
+                }
+              }
+            }
+          }
+        } catch (error) {
+          console.error('Error getting RPC URL:', error);
+        }
+        return 'unknown';
```
small nit: It should be safe to trust the return types here. If the type indicates that value will not return `undefined`/`null`/etc you should be able to trust it (if you found that we can't trust it, either the code or the types are broken and should be fixed or a new issue created to fix them).

We _probably_ don't need all of the falsey checks (`if (networkClientId)`, `if (networkConfig)`), , but I haven't checked all of these return types myself.

There are cases when we must should guard against unexpected data (like we do in migration files, and to a much larger degree), I'm not sure if this is one of those cases (I could be wrong though!)

httpJunkie replied:
ok, I've updated I will do some more tests now to ensure everything is working. But looking at the code closer and types. I think it's safe.

httpJunkie committed:
test: update metrics.test.ts to expect `private` rather than `unknown`

httpJunkie committed:
fix: fetch-with-cache import fixed and now using a type instead of an interface

httpJunkie commented on app/scripts/lib/transaction/metrics.test.ts:
```diff
@@ -819,6 +819,9 @@ describe('Transaction metrics', () => {
         is_smart_transaction: undefined,
         transaction_advanced_view: undefined,
+        // Add the new properties that our implementation adds
+        hd_entropy_index: undefined,
```
I dont know how those comments got there, it's weird because I don't remember adding that hd_entropy_index... field. But I do remember a merge conflict that had that field.  But unless it's showing a change because the order is different from main. I don't know but I will investigate.

I will remove them.

httpJunkie replied:
The only line that should have been added here is:
```
rpc_domain: 'unknown',
```

And I am removing anything else. I'm baffled. Maybe it was a copy paste mistake but they are not needed in the test but it was still passing. Maybe that is why i didn't notice. Maybe I was trying to figure out what I needed to make the test pass. But it really makes no sense. Sorry about that.

davidmurdoch replied:
No problem. I always do a thorough self-review of every single line of my own pull requests. Its helpful to catch things like this before a review spends time trying to understand why something has been added changed that might have crept in for all sort of various reasons. I also leave comments attempting to explain some of the more complicated changes and try to explain why I might have chosen one way of doing something over another.

Here's a recent example: https://github.com/MetaMask/metamask-extension/pull/31404

Most of these comments are my own, even though it is my own PR.

![image](https://github.com/user-attachments/assets/29ddb5df-a72f-4b6c-8481-e7b29ef4c60b)

httpJunkie committed:
test: update metrics test to only include properties needed for the test considering changes made for `rpc_domain`

httpJunkie commented on app/scripts/lib/transaction/metrics.ts:
```diff
@@ -1098,6 +1099,20 @@ async function buildEventFragmentProperties({
 
+  // Get RPC URL from provider
+  let rpcUrl = 'invalid';
+  if (transactionMetricsRequest.getNetworkRpcUrl && transactionMeta.chainId) {
```
This is a TS without. BUT, if we make `getNetworkRpcUrl` required in the type definition, we'd still need similar conditional logic to check if `transactionMeta.chainId` exists before calling the function.

Making `getNetworkRpcUrl` required would be a breaking change and (could) force updates to implementations of `TransactionMetricsRequest`, (like tests). Wouldn't we need tp identify everywhere type is used and ensure the property is implemented? especially where `getNetworkRpcUrl` isn't relevant?

I added it because I got: "Cannot invoke an object which is possibly `undefined`"

But looking into it more and the path where we remove it and make a few changes seems like the less ideal solution. But, if I'm missing something, LMK.

davidmurdoch replied:
> Making `getNetworkRpcUrl` required would be a breaking change

Why do you say that? You are adding the `getNetworkRpcUrl ` method in this PR, so you can define it however you'd like. the only consumer of the `TransactionMetricsRequest` type is in this repo, so you don't have to worry about breaking any thing else.

> Wouldn't we need tp identify everywhere type is used

That is what the type system is supposed to do for us. Plus all the unit and e2e tests that are being run.

> I added it because I got: "Cannot invoke an object which is possibly undefined"

I think you'll get `Cannot invoke an object which is possibly undefined` only if you keep the type definition for `getNetworkRpcUrl` as optional (`getNetworkRpcUrl?: (chainId: string) => string;` vs `getNetworkRpcUrl: (chainId: string) => string;`)

And for `chainId`, our types indicate that it will always be defined. Are you finding that not to be true?

httpJunkie commented on app/scripts/lib/transaction/metrics.ts:
```diff
@@ -1098,6 +1099,20 @@ async function buildEventFragmentProperties({
     transactionMetricsRequest.getAccountBalance,
   );
 
+  // Get RPC URL from provider
+  let rpcUrl = 'invalid';
+  if (transactionMetricsRequest.getNetworkRpcUrl && transactionMeta.chainId) {
+    rpcUrl = transactionMetricsRequest.getNetworkRpcUrl(
+      transactionMeta.chainId,
+    );
+  }
+  const domain = extractRpcDomain(rpcUrl);
+
+  // Only add domain to properties for "Transaction Submitted" and "Transaction Finalized" events
```
Good eye, this code changed quite a bit. I have moved the domain calculation logic inside the conditional.

httpJunkie committed:
fix:  moved the domain calculation logic inside the conditional

httpJunkie committed:
refactor: remove template URL cleaning from RPC provider domains

Remove unnecessary URL template variable cleaning step from initializeRpcProviderDomains
and update corresponding tests. The URL constructor handles template variables without
requiring explicit cleaning, simplifying the code while maintaining functionality.

The changes include:
- Remove template variable replacement in initializeRpcProviderDomains
- Remove template-specific tests from RPC URL handling utilities tests
- Maintain core domain initialization and validation functionality

httpJunkie commented on app/scripts/lib/util.ts:
```diff
@@ -480,3 +481,138 @@ export function getConversionRatesForNativeAsset({
+          for (const rpcUrl of chain.rpc) {
+            try {
+              // Handle placeholder variables like ${INFURA_API_KEY}
+              const cleanedUrl = rpcUrl.replace(/\$\{[^}]+\}/gu, 'API_KEY');
```
Removed. And added tests for these functions. I thought for sure I could create a test and show you. But I was wrong.

httpJunkie committed:
refactor: simplify RPC domain extraction logic

The RPC domain extraction logic has been simplified to only differentiate
between known domains and private/invalid URLs, removing special case handling
for different URL types.

- Simplified extractRpcDomain to return known domain or 'private'/'invalid'
- Removed special handling for localhost, IPv4/IPv6, punycode domains
- Consolidated and simplified test cases to match new implementation
- Kept URL parsing and error handling for invalid URLs

This change reduces complexity while maintaining the core functionality of
identifying known RPC providers versus private/custom networks.

httpJunkie committed:
fix: removing special case handling
for different URL types.

httpJunkie commented on app/scripts/lib/transaction/metrics.test.ts:
```diff
@@ -819,6 +819,9 @@ describe('Transaction metrics', () => {
         ],
         is_smart_transaction: undefined,
         transaction_advanced_view: undefined,
+        // Add the new properties that our implementation adds
```
It's a mistake.

httpJunkie committed:
refactor: simplify network RPC URL retrieval logic

The network RPC URL retrieval logic has been simplified by trusting the type
system and removing redundant null checks.

- Removed unnecessary checks for networkClientId and networkConfig existence
- Combined endpoint URL checks into a single return statement
- Maintained error handling for external method calls
- Kept fallback to 'unknown' for error cases

httpJunkie committed:
Merge branch 'main' into feat/metrics-for-rpc-endpoints

httpJunkie committed:
fix: add semi

httpJunkie committed:
Merge branch 'main' into feat/metrics-for-rpc-endpoints

httpJunkie committed:
Merge branch 'main' into feat/metrics-for-rpc-endpoints

httpJunkie committed:
Merge branch 'main' into feat/metrics-for-rpc-endpoints

httpJunkie commented on app/scripts/lib/transaction/metrics.ts:
```diff
@@ -1098,6 +1099,19 @@ async function buildEventFragmentProperties({
   );
 
+  // Only calculate and add domain to properties for "Transaction Submitted" and "Transaction Finalized" events
+  if (status === 'submitted' || status === 'confirmed') {
```
I can add those. I don’t have approvals yet from David so I can get those changes made now.

httpJunkie committed:
fix: use TransactionStatus.confirmed and TransactionStatus.submitted instead of strings

httpJunkie committed:
Merge branch 'feat/metrics-for-rpc-endpoints' of github.com:MetaMask/metamask-extension into feat/metrics-for-rpc-endpoints

httpJunkie committed:
fix: typo

httpJunkie committed:
Merge branch 'main' into feat/metrics-for-rpc-endpoints

vinistevam reviewed (dismissed):
All good from the Confirmations side. :+1:

httpJunkie committed:
Merge branch 'main' into feat/metrics-for-rpc-endpoints

davidmurdoch commented on app/scripts/lib/util.ts:
```diff
@@ -480,3 +481,97 @@ export function getConversionRatesForNativeAsset({
+        if (chain.rpc && Array.isArray(chain.rpc)) {
+          for (const rpcUrl of chain.rpc) {
+            try {
+              // Removed cleaning step - will fail for URLs like:
+              // https://${INFURA_API_KEY}.infura.io/v3/
```
This comment looks like it might be a reply to my previous PR review comment? I don't think it should be in the code.

That said,
`new URL("https://{INFURA_API_KEY}.infura.io/v3/").hostname` returns `${infura_api_key}.infura.io`. This seems like _correct_ behavior to me.

davidmurdoch commented on app/scripts/lib/util.ts:
```diff
@@ -480,3 +481,97 @@ export function getConversionRatesForNativeAsset({
+    try {
+      url = new URL(rpcUrl);
+    } catch (e) {
+      if (!rpcUrl.startsWith('http://') && !rpcUrl.startsWith('https://')) {
```
I don't think this catch is needed either. If the `rpcUrl` is not a valid URL the rest of the application won't work anyway.

httpJunkie commented on app/scripts/lib/util.ts:
```diff
@@ -480,3 +481,97 @@ export function getConversionRatesForNativeAsset({
+ * @param domain - The domain to check
+ */
+export function isKnownDomain(domain: string): boolean {
+  return knownDomainsSet?.has(domain) ?? false;
```
I can make this change. It should have no other impact and since domain names are case-insensitive by internet standards, this change would make our implementation more accurate.

httpJunkie replied:
Good catch! You're right - domains should be case-insensitive. I'll update both places to use lowercase for consistency. This aligns better w/ standards should prevent same domain with different capitalization from appearing as separate entries in our metrics. Test was expecting this behavior all along, so it's a nice cleanup that makes the code better.

httpJunkie committed:
fix: ensure case-insensitive domain name handling

Make domain name comparisons case-insensitive by converting to lowercase
during both initialization and lookup. This matches real-world DNS
behavior and prevents metric fragmentation.

httpJunkie committed:
update link in changelog

httpJunkie committed:
remove comment

davidmurdoch commented on app/scripts/lib/util.ts:
```diff
@@ -480,3 +481,97 @@ export function getConversionRatesForNativeAsset({
+              // Removed cleaning step - will fail for URLs like:
+              // https://${INFURA_API_KEY}.infura.io/v3/
+              const url = new URL(rpcUrl);
+              knownDomainsSet.add(url.hostname.toLowerCase());
```
The `hostname` property will always be normalized to lowercase by `URL`, per the URL standard.
```suggestion
              knownDomainsSet.add(url.hostname);
```

httpJunkie committed:
fix: improve RPC domain extraction with protocol handling

- Add automatic protocol prefix for URLs without http(s)://
- Fix known domains testing support in extractRpcDomain
- Update test expectations to match actual behavior
- Handle both full URLs and domain-only formats

Fixes failing tests in util.test.js for extractRpcDomain function

davidmurdoch commented on app/scripts/lib/util.ts:
```diff
@@ -480,3 +481,97 @@ export function getConversionRatesForNativeAsset({
+ * @param domain - The domain to check
+ */
+export function isKnownDomain(domain: string): boolean {
+  return knownDomainsSet?.has(domain?.toLowerCase()) ?? false;
```
`domain` should always be defined here, per the type definition.

```suggestion
  return knownDomainsSet?.has(domain?.toLowerCase()) ?? false;
```

httpJunkie committed:
Update app/scripts/lib/util.ts

Co-authored-by: David Murdoch <187813+davidmurdoch@users.noreply.github.com>

httpJunkie committed:
fix: update getNetworkRpcUrl chainId parameter type to Hex

The chainId parameter type in getNetworkRpcUrl was incorrectly typed as
string. Updated to use the Hex type to match the actual chainId values
being passed from transactionMeta.chainId.

update test formatting

httpJunkie committed:
test: add getNetworkRpcUrl mock to transaction metrics tests

Added missing getNetworkRpcUrl mock function to test objects in both
metrics.test.ts and metametrics.test.ts to match the updated
TransactionMetricsRequest type definition.

httpJunkie committed:
test: update util.test.js

Moves the 'invalid-url-format' test to the "handle invalid URLs and edge cases" block where it logically belongs
Keeps all 'private' domain tests in their own block
Maintains consistent formatting

httpJunkie committed:
test: restore knownDomainsForTesting parameter for extractRpcDomain

Added knownDomainsForTesting parameter to allow proper testing of the
domain extraction logic without relying on mocked internal functions.
This enables direct testing of URLs with known domains while keeping
the production code path unchanged. Updated tests to use this parameter
and to verify both protocol and protocol-less URL handling.

httpJunkie committed:
fix: add Jest mock type assertion for getNetworkRpcUrl

The fix adds proper type assertion for Jest mock function to resolve TypeScript
error where 'mockReturnValue' was not recognized on the function type.

Issue: Property 'mockReturnValue' does not exist on type
'(chainId: `0x${string}`) => string'

httpJunkie committed:
Merge branch 'main' into feat/metrics-for-rpc-endpoints

httpJunkie committed:
Merge branch 'main' into feat/metrics-for-rpc-endpoints

httpJunkie committed:
Merge branch 'main' into feat/metrics-for-rpc-endpoints

httpJunkie committed:
Merge branch 'main' into feat/metrics-for-rpc-endpoints

httpJunkie committed:
Merge branch 'main' into feat/metrics-for-rpc-endpoints

httpJunkie committed:
Merge branch 'main' into feat/metrics-for-rpc-endpoints

httpJunkie committed:
Merge branch 'main' into feat/metrics-for-rpc-endpoints

httpJunkie committed:
Merge branch 'main' into feat/metrics-for-rpc-endpoints

davidmurdoch commented:
blocked by https://github.com/MetaMask/metamask-extension/pull/32297

httpJunkie committed:
Merge branch 'main' into feat/metrics-for-rpc-endpoints

httpJunkie committed:
Merge branch 'main' into feat/metrics-for-rpc-endpoints

davidmurdoch committed:
use cached data only

httpJunkie committed:
Merge branch 'main' into feat/metrics-for-rpc-endpoints

httpJunkie committed:
Merge branch 'main' into feat/metrics-for-rpc-endpoints
